### PR TITLE
Use global and local setup/terdown when using chip::Test::AppContext

### PR DESCRIPTION
**Pull request by arkq**
This PR is a followup for #30548

### Changes

- create convenient `SetUpTestSuite`/`TearDownTestSuite` `SetUp`/`TearDown` wrappers for nlTest in `chip::Test::AppContext`
- use global setup/terdown and per test setup/teardown in all places where `chip::Test::AppContext` is used
- fix test order dependencies in `TestEventLogging` and `TestEventLoggingNoUTCTime` (counters in asserts has been tuned for running tests in defined order)

### Testing

Tested with patches provided in #30548 and #30470 (so after transition to pigweed test, everything should be OK)
CI will check for potential failures

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/app/tests/AppTestContext.cpp [hunk 1]
```diff
@@ -38,28 +38,48 @@ chip::Access::AccessControl gPermissiveAccessControl;
...
 
-CHIP_ERROR AppContext::Init()
+CHIP_ERROR AppContext::SetUpTestSuite()
 {
-    ReturnErrorOnFailure(Super::Init());
-    ReturnErrorOnFailure(chip::DeviceLayer::PlatformMgr().InitChipStack());
-    ReturnErrorOnFailure(chip::app::InteractionModelEngine::GetInstance()->Init(&GetExchangeManager(), &GetFabricTable(),
-                                                                                app::reporting::GetDefaultReportScheduler()));
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit((err = chip::Platform::MemoryInit()) == CHIP_NO_ERROR,
```
[the remark this answers is not in the document]
Comment (arkq) [reply src/app/tests/AppTestContext.cpp:44]: It calls `Init()` of the super but in an explicit way by calling `LoopbackTransportManager::Init()` (I forgot to remove ` typedef LoopbackMessagingContext Super;` from the header).

I can change that to `Super::Init()`, but IMHO it obfuscates the setup logic.

File: src/app/tests/TestICDManager.cpp [hunk 2]
```diff
@@ -319,28 +317,21 @@ class TestICDManager
...
-{
-    NL_TEST_DEF("TestICDModeDurations",         TestICDManager::TestICDModeDurations),
-    NL_TEST_DEF("TestKeepActivemodeRequests",   TestICDManager::TestKeepActivemodeRequests),
-    NL_TEST_DEF("TestICDMRegisterUnregisterEvents", TestICDManager::TestICDMRegisterUnregisterEvents),
-    NL_TEST_SENTINEL()
+static const nlTest sTests[] = {
+    NL_TEST_DEF("TestICDModeDurations", TestICDManager::TestICDModeDurations),
+    // NL_TEST_DEF("TestKeepActivemodeRequests",   TestICDManager::TestKeepActivemodeRequests),
+    // NL_TEST_DEF("TestICDMRegisterUnregisterEvents", TestICDManager::TestICDMRegisterUnregisterEvents),
+    NL_TEST_SENTINEL(),
```
Comment (mbknust) [inline src/app/tests/TestICDManager.cpp:324]: Why are these commented out? Could it be a mistake?

Comment (arkq) [reply src/app/tests/TestICDManager.cpp:324]: Thanks for spotting that! Fixed

File: src/app/tests/AppTestContext.h [hunk 3]
```diff
@@ -26,14 +26,41 @@ namespace Test {
...
-
 public:
-    /// Initialize the underlying layers.
-    CHIP_ERROR Init() override;
+    // Performs shared setup for all tests in the test suite
+    virtual CHIP_ERROR SetUpTestSuite();
+    // Performs shared teardown for all tests in the test suite
+    virtual void TearDownTestSuite();
+    // Performs setup for each individual test in the test suite
+    virtual CHIP_ERROR SetUp();
```
Comment (cecille) [inline src/app/tests/AppTestContext.h:35]: General Q, not a criticism of this PR.

Is this part of the swap over to the pw unit test from nlUnitTest? If so, does the pw unit test framework support error returns on setup functions?

**Review by cecille — APPROVED** [review]

File: src/app/tests/AppTestContext.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/tests/AppTestContext.cpp:44]: I think LoopbackMessagingContext is not the same as LoopbackTransportManager.

[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/tests/AppTestContext.cpp:44]: I thought this was resolved however I did indeed not read it through :(

Reading thgouth it: LoopbackMessagingContext::Init does things doday, specifically memory init, loppbacktransportmanager init AND messagingcontext init.

New update copies over only the first 2 and does not init the messaging context, but initializes the chip stack instead.

It seems slightly unusual for a class init to not call the base class init. If this is done on purpose it should explain why (in this case why is it undesirable to do MessagingContext::init?

**Pull request merged**